mitiq/rem/inverse_confusion_matrix.py

```diff
@@ -0,0 +1,108 @@
+# Copyright (C) 2022 Unitary Fund
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+from typing import List
+import numpy as np
+import numpy.typing as npt
+
+from mitiq._typing import MeasurementResult, Bitstring
+
+
+def sample_probability_vector(
+    probability_vector: npt.NDArray[np.float64], samples: int
+) -> npt.NDArray[np.uint8]:
+    """Generate a number of samples from a probability distribution as
+    bitstrings.
+
+    Args:
+        probability_vector: A probability vector.
+    """
+    # sample using the probability distribution given
+    num_values = len(probability_vector)
+    choices = np.random.choice(num_values, size=samples, p=probability_vector)
+
+    # convert samples to binary strings
+    bit_width = int(np.log2(num_values))
+    binary_repr_vec = np.vectorize(np.binary_repr)
+    binary_strings = binary_repr_vec(choices, width=bit_width)
+
+    # split the binary strings into an array of ints
+    bitstrings = np.apply_along_axis(
+        np.fromstring, 1, binary_strings[:, None], dtype="U1"
+    ).astype(np.uint8)
+
+    return bitstrings
+
+
+def bitstrings_to_probability_vector(
+    bitstrings: List[Bitstring],
+) -> npt.NDArray[np.float64]:
+    """Converts a list of measured bitstrings to a probability vector estimated
+    as the empirical frequency of each bitstring (ordered with increasing
+    binary value).
+
+    Args:
+        bitstrings: All measured bitstrings.
+    """
+    pv = np.zeros(2 ** len(bitstrings[0]))
+    for bs in bitstrings:
+        index = int("".join([str(bit) for bit in bs]), base=2)
```

Comment:
I _think_ this should be marginally faster, and IMO a bit cleaner, but it's also unlikely to make much of an effect since the bitstrings we are likely to deal with are relatively short. Take it or leave it!
```suggestion
        index = int("".join(map(str, bs)), base=2)
```

Reply:
I like that much better. Thanks.